MOD renamed `axis` meta-Data field to `orientation

AndreasThenn opened the pull request:
related to #15

AndreasThenn committed:
MOD renamed `axis` meta-Data field to `orientation

ktraunmueller commented on Source/Components/StackCluster.swift:
```diff
@@ -35,11 +35,15 @@ extension ClusterLayout {
             let backgroundColor = hexString.flatMap { UIColor(hexString: $0) }
-
+            
+            if "horizontal" == orientationRawValue {
```
We should maybe introduce an enum with (`String`-valued) raw values for this and use the failable initializer to map strings onto enumeration cases.

Something like this:
```
var axis = UILayoutConstraintAxis.vertical
if let orientationRawValue = meta["orientation"] as? String,
   let orientation = Orientation(rawValue: orientationRawValue) {
     axis = orientation.layoutConstraintAxis
}
```
with
```
public enum ClusterLayout {
    
    public enum Orientation: String {
        case horizontal
        case vertical
    }
}

extension ClusterLayout.Orientation {
    
    var layoutConstraintAxis: UILayoutConstraintAxis {
        switch self {
        case .horizontal: return .horizontal
        case .vertical: return .vertical
        }
    }
}
```

AndreasThenn replied:
I like your approach, I tried to get the String name of out of UILayoutConstraintAxis, but that did not work. I updated the PR

AndreasThenn committed:
CRC

AndreasThenn committed:
CRC

ChristianOrgler commented on Source/Components/StackCluster.swift:
```diff
@@ -35,11 +35,14 @@ extension ClusterLayout {
-            let axisRawValue = meta["axis"] as? Int
-            let axis = axisRawValue.flatMap { UILayoutConstraintAxis(rawValue: $0) }
+            var axis = UILayoutConstraintAxis.vertical
+            if let orientationRawValue = meta["orientation"] as? String,
```
don't you need to adapt the example json as well?

AndreasThenn replied:
It already used `orientation`

ChristianOrgler reviewed (approved):
Please adapt any example .json files from axis to orientation.

joanromano commented on Source/Components/ClusterLayout.swift:
```diff
@@ -14,5 +14,17 @@ import Foundation
+    public enum Orientation: String {
+        case horizontal
+        case vertical
+    }
```
I am not sure if this information belongs here: afaik the `ClusterLayout` public enum is an enumeration to add new clusters, and the orientation is something that is not used in all of them.

Plus, I don't find this enum useful to be exposed publicly. When would this be used? As it is right now is only used by the library to parse the `ComponentMeta` into a `ExpressibleByComponentMeta` stack config. I would keep this thing internal to the library somewhere else if you really need to use it (which maybe we don't really need it for now).

ChristianOrgler replied:
I agree 👍

ktraunmueller replied:
Right, good point.

joanromano commented on Source/Components/StackCluster.swift:
```diff
@@ -35,11 +35,14 @@ extension ClusterLayout {
+            if let orientationRawValue = meta["orientation"] as? String,
+                let orientation = Orientation(rawValue: orientationRawValue) {
+                axis = orientation.layoutConstraintAxis
+            }
```
If it's only for this, I don't think a `Orientation` enum is helpful for now. I would remove it and do the parsing normally as with the other fields

MP0w replied:
+1 is not something helpful for other Components

ktraunmueller replied:
Agreed, no need to expose that. Maybe make it a private type (of `StackCluster`)?

joanromano replied:
If you really think it's useful to keep it for the parsing ok, yes I would keep it private.

AndreasThenn replied:
thx for the feedback, @joanromano I updated my PR

joanromano replied:
Thanks, please don't forget to update the tests and the JSON schema guide as mentioned above.

joanromano reviewed (changes requested):
Please update the `StackClusterTests` where using a dictionary (`ComponentMeta`); those tests are still using `axis` but they pass because .vertical is now the default value when parsing. Adding a new test with this new `orientation` String value would have helped here. And I would really rethink about the need of that new `Orientation` enum.

But more importantly, please update the JSON schema guide https://github.com/runtastic/Matrioska/blob/master/Documentation/JSON%20schema%20guide.md otherwise it's reflecting wrong information.

AndreasThenn committed:
CRC

AndreasThenn committed:
MOD updated documentation

joanromano reviewed (changes requested):
Please update `StackClusterTests`, I believe there are tests using `axis` in their `ComponentMeta` and they pass just because we test against the default value.

AndreasThenn committed:
MOD changed StackClusterTests when trying to read from dictionary to use orientation instead of axis now.
ADD new test to check for vertical orientation, to not use default orientation

AndreasThenn commented:
@joanromano thx, indeed one test just passed because it used default value. I modified it to use orientation and added a new one to not use the default value. I added a new one, because the horizontal orientation of the first one just looked really ugly. Please check if you're fine with the new setup.
